packages/scenario/test/repeat.test.ts

```diff
@@ -2694,3 +2694,429 @@ describe('Tests ported from JavaRosa - repeats', () => {
+			scenario.answer('/data/rep-count', 4);
+
+			expect(scenario.answerOf('/data/rep[1]/calc')).toEqualAnswer(intAnswer(2));
+			expect(scenario.answerOf('/data/rep[2]/calc')).toEqualAnswer(intAnswer(4));
+			expect(scenario.answerOf('/data/rep[3]/calc')).toEqualAnswer(intAnswer(6));
+			expect(scenario.answerOf('/data/rep[4]/calc')).toEqualAnswer(intAnswer(8));
+		});
+	});
+
+	describe('control of count by engine', () => {
+		let scenario: Scenario;
+
+		beforeEach(async () => {
+			scenario = await Scenario.init(
+				'Direct reference count updates',
+				// prettier-ignore
+				html(
+					head(
+						title('Direct reference count updates'),
+						model(
+							mainInstance(
+								t(
+									'data id="direct-reference-count-updates"',
+									t('rep-count', '2'),
+									t('rep',
+										t('quest')))
+							),
+							bind('/data/rep-count'),
+							bind('/data/rep/quest').type('string')
+						)
+					),
+					body(
+						input('/data/rep-count'),
+						repeat('/data/rep', '/data/rep-count',
+							input('/data/rep/quest')
+						)
+					)
+				)
+			);
+		});
+
+		it('does not allow a client (user) to add or remove instances of a count-controlled repeat', () => {
+			expect(scenario.proposed_canCreateOrRemoveRepeat('/data/rep')).toBe(false);
+		});
+	});
+});
+
+describe('jr:noAddRemove', () => {
+	let scenario: Scenario;
+
+	beforeEach(async () => {
+		scenario = await Scenario.init(
+			'Direct reference count updates',
+			// prettier-ignore
+			html(
+				head(
+					title('Direct reference count updates'),
+					model(
+						mainInstance(
+							t(
+								'data id="direct-reference-count-updates"',
+								t('rep1',
+									t('quest')),
+								t('rep1',
+									t('quest')),
+								t('rep2 jr:template=""',
+									t('quest')),
+								t('rep2',
+									t('quest')),
+								t('rep2',
+									t('quest')),
+								t('rep2',
+									t('quest')),
+								t('rep3 jr:template=""',
+									t('quest')))
+						),
+						bind('/data/rep1/quest').type('string'),
+						bind('/data/rep2/quest').type('string'),
+						bind('/data/rep3/quest').type('string')
+					)
+				),
+				body(
+					t('repeat nodeset="/data/rep1" jr:noAddRemove="true()"',
+						input('/data/rep1/quest')
+					),
+					t('repeat nodeset="/data/rep2" jr:noAddRemove="true()"',
+						input('/data/rep2/quest')
+					),
+					t('repeat nodeset="/data/rep3" jr:noAddRemove="true()"',
+						input('/data/rep3/quest')
+					)
+				)
+			)
+		);
+	});
+
+	it('creates the number of repeats defined in the form', () => {
+		expect(scenario.countRepeatInstancesOf('/data/rep1')).toBe(2);
+	});
+
+	it('does not create a repeat instance for a template where the form defines subsequent repeat instances', () => {
+		expect(scenario.countRepeatInstancesOf('/data/rep2')).toBe(3);
+	});
+
+	// Subjective! Maybe a form design mistake? But it wouldn't ever make sense to
+	// create a fixed set of zero repeat instances.
```

Comment:
I agree with this.

Reply:
I think we can consider producing a warning about this (and things like it) in the future.